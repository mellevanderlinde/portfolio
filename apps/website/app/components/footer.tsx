import type { ComponentType, ReactNode } from 'react'
import { FaCode, FaGithub, FaLinkedinIn } from 'react-icons/fa6'
import { name } from './nav'

const YEAR = new Date().getFullYear()

const links = {
  github: 'https://github.com/mellevanderlinde',
  linkedin: 'https://linkedin.com/in/mellevdlinde',
  source: 'https://github.com/mellevanderlinde/portfolio',
}

interface LinkProps {
  href: string
  icon: ComponentType
}

function Link({ href, icon: Icon }: LinkProps): ReactNode {
  return (
    <a href={href} target="_blank" rel="noopener noreferrer">
      <Icon />
    </a>
  )
}

function Links(): ReactNode {
  return (
    <div className="flex text-lg gap-3.5 float-right transition-opacity duration-300 hover:opacity-90">
      <Link href={links.github} icon={FaGithub} />
      <Link href={links.linkedin} icon={FaLinkedinIn} />
      <Link href={links.source} icon={FaCode} />
    </div>
  )
}

export default function Footer(): ReactNode {
  return (
    <small className="block lg:mt-24 mt-16 text-[#1C1C1C] dark:text-[#D4D4D4]">
      <time>
        Â©
        {' '}
        {YEAR}
      </time>
      {' '}
      {name}
      <Links />
    </small>
  )
}
