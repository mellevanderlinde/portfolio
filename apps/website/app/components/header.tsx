import type { ReactNode } from "react";

export function Header1({ title }: { title: string }): ReactNode {
  return <h1 className="mb-8 text-2xl font-medium tracking-tight">{title}</h1>;
}
