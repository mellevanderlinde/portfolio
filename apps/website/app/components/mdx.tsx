import type { MDXRemoteProps } from 'next-mdx-remote/rsc'
import type { ReactNode } from 'react'
import hljs from 'highlight.js/lib/core'
import typescript from 'highlight.js/lib/languages/typescript'
import yaml from 'highlight.js/lib/languages/yaml'
import { MDXRemote } from 'next-mdx-remote/rsc'
import Link from 'next/link'
import { createElement } from 'react'

hljs.registerLanguage('typescript', typescript)
hljs.registerLanguage('yaml', yaml)

function CustomLink(props: { href: string, children: ReactNode }): ReactNode {
  const href = props.href
  if (href.startsWith('/')) {
    return <Link {...props}>{props.children}</Link>
  }
  if (href.startsWith('#')) {
    return <a {...props} />
  }
  return <a target="_blank" rel="noopener noreferrer" {...props} />
}

function Code({
  children,
  ...props
}: { children: string, className?: string }): ReactNode {
  const language = props.className?.replace('language-', '') || 'yaml'
  const __html = hljs.highlight(children, { language }).value
  return <code dangerouslySetInnerHTML={{ __html }} {...props} />
}

function Callout(props: {
  emoji: ReactNode
  children: ReactNode
}): ReactNode {
  return (
    <div className="px-4 py-3 rounded p-1 text-sm flex items-center text-neutral-900 dark:text-neutral-100 mb-8 bg-white dark:bg-[#0f1113] text-black border border-neutral-200 dark:border-neutral-700">
      <div className="flex items-center w-4 mr-4">{props.emoji}</div>
      <div className="w-full callout leading-relaxed">{props.children}</div>
    </div>
  )
}

function slugify(str: string): string {
  return str
    .toString()
    .toLowerCase()
    .trim()
    .replace(/\s+/g, '-')
    .replace(/&/g, '-and-')
    .replace(/[^\w-]+/g, '')
    .replace(/-{2,}/g, '-')
}

function createHeading(level: number) {
  const Heading = ({ children }: { children: string }) => {
    const slug = slugify(children)
    return createElement(
      `h${level}`,
      { id: slug },
      [
        createElement('a', {
          href: `#${slug}`,
          key: `link-${slug}`,
          className: 'anchor',
        }),
      ],
      children,
    )
  }
  Heading.displayName = `Heading${level}`
  return Heading
}

const components = {
  h1: createHeading(1),
  h2: createHeading(2),
  h3: createHeading(3),
  h4: createHeading(4),
  h5: createHeading(5),
  h6: createHeading(6),
  a: CustomLink,
  code: Code,
  Callout,
}

export function CustomMDX(props: MDXRemoteProps): ReactNode {
  return (
    <MDXRemote
      {...props}
      components={{ ...components, ...(props.components || {}) }}
    />
  )
}
