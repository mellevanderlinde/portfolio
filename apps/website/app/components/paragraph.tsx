import { ReactNode } from "react";

export function Paragraph({ content }): ReactNode {
  return (
    <p className="mb-2 prose prose-neutral dark:prose-invert">{content}</p>
  );
}
