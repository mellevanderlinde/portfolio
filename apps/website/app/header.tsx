'use client'
import type { ReactNode } from 'react'
import Link from 'next/link'
import { TextEffect } from '@/components/ui/text-effect'

export function Header(): ReactNode {
  return (
    <header className="mb-8 flex items-center justify-between">
      <div>
        <Link href="/" className="font-medium text-black dark:text-white">
          Melle van der Linde
        </Link>
        <TextEffect
          as="p"
          preset="fade"
          per="char"
          className="text-zinc-600 dark:text-zinc-500"
          delay={0.5}
        >
          Cloud Engineer
        </TextEffect>
      </div>
    </header>
  )
}
