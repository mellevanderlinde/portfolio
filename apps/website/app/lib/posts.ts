import fs from 'node:fs'
import path from 'node:path'

interface Metadata {
  title: string
  publishedAt: string
}

interface BlogPost {
  metadata: Metadata
  slug: string
  content: string
}

function parseFrontmatter(fileContent: string): Omit<BlogPost, 'slug'> {
  // eslint-disable-next-line regexp/no-super-linear-backtracking
  const frontmatterRegex = /---\s*([\s\S]*?)\s*---/
  const match = frontmatterRegex.exec(fileContent)
  if (!match) {
    throw new Error('Frontmatter not found')
  }
  const frontMatterBlock = match[1]
  const content = fileContent.replace(frontmatterRegex, '').trim()
  const frontMatterLines = frontMatterBlock.trim().split('\n')
  const metadata: Partial<Metadata> = {}

  for (const line of frontMatterLines) {
    const [key, ...valueArr] = line.split(': ')
    let value = valueArr.join(': ').trim()
    value = value.replace(/^['"](.*)['"]$/, '$1')
    metadata[key.trim() as keyof Metadata] = value
  }

  return { metadata: metadata as Metadata, content }
}

function getMDXFiles(dir: string): string[] {
  return fs.readdirSync(dir).filter(file => path.extname(file) === '.mdx')
}

function readMDXFile(filePath: string): Omit<BlogPost, 'slug'> {
  const rawContent = fs.readFileSync(filePath, 'utf-8')
  return parseFrontmatter(rawContent)
}

function getMDXData(dir: string): BlogPost[] {
  const mdxFiles = getMDXFiles(dir)
  return mdxFiles.map((file) => {
    const { metadata, content } = readMDXFile(path.join(dir, file))
    const slug = path.basename(file, path.extname(file))
    return {
      metadata,
      slug,
      content,
    }
  })
}

export function getBlogPosts(): BlogPost[] {
  // eslint-disable-next-line node/prefer-global/process
  return getMDXData(path.join(process.cwd(), 'content'))
}

export function formatDate(date: string): string {
  return new Date(date).toLocaleString('en-us', {
    month: 'short',
    day: 'numeric',
    year: 'numeric',
  })
}
