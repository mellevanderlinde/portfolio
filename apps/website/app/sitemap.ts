import type { MetadataRoute } from 'next'
import { url } from '@/lib/constants'
import { posts } from '../lib/data'

export const dynamic = 'force-static'

export default function sitemap(): MetadataRoute.Sitemap {
  const lastModified = new Date()
  lastModified.setUTCDate(lastModified.getUTCDate())
  lastModified.setUTCHours(11, 0, 0, 0)

  return [
    { url, lastModified },
    ...posts.map(post => ({ url: `${url}${post.link}`, lastModified })),
  ]
}
