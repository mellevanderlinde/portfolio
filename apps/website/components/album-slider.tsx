'use client'

import type { ReactNode } from 'react'
import { InfiniteSlider } from '@/components/motion-primitives/infinite-slider'
import { UnderlinedLink } from '@/components/underlined-link'
import { transition, variantsContainer, variantsSection } from '@/lib/constants'
import { albums } from '@/lib/data'
import { motion } from 'motion/react'

function InfiniteAlbumSlider(): ReactNode {
  return (
    <InfiniteSlider speedOnHover={20} gap={24} className="mt-8">
      {albums.sort(() => Math.random() - 0.5) // Randomly shuffle albums
        .map(album => (
          <img
            key={album.id}
            src={`https://i.scdn.co/image/${album.id}`}
            alt={album.name}
            className="aspect-square w-[120px] rounded-[4px]"
          />
        ))}
    </InfiniteSlider>
  )
}

export function AlbumSlider(): ReactNode {
  return (
    <motion.main
      className="space-y-24"
      variants={variantsContainer}
      initial="hidden"
      animate="visible"
    >
      <motion.section
        variants={variantsSection}
        transition={transition}
        className="mt-18"
      >
        <div className="flex-1">
          <p className="text-zinc-600 dark:text-zinc-400">
            These are a few of my favorite albums. You can also find me on
            {' '}
            <UnderlinedLink href="https://soundcloud.com/melledj">SoundCloud</UnderlinedLink>
            .
          </p>
        </div>
        <InfiniteAlbumSlider />
      </motion.section>
    </motion.main>
  )
}
