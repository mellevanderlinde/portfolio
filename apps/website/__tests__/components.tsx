import { render } from "@testing-library/react";
import { Header1 } from "@app/components/header";
import { Navbar } from "@app/components/nav";
import { getBlogPosts } from "@app/lib/posts";
import Footer from "@app/components/footer";
import robots from "@app/robots";
import sitemap from "@app/sitemap";

test("Match footer", () => {
  const { container } = render(<Footer />);
  expect(container).toMatchSnapshot();
});

test("Match header", () => {
  const { container } = render(<Header1 title="Test" />);
  expect(container).toMatchSnapshot();
});

test("Match navbar", () => {
  const { container } = render(<Navbar />);
  expect(container).toMatchSnapshot();
});

test("Match posts", () => {
  const result = getBlogPosts();
  expect(result).toMatchSnapshot();
});

test("Match robots", () => {
  const result = robots();
  expect(result).toMatchSnapshot();
});

test("Match sitemap", () => {
  const result = sitemap();
  // Change to static date for snapshot
  result.forEach((item) => {
    item.lastModified = "2024-01-01T00:00:00.000Z";
  });
  expect(result).toMatchSnapshot();
});
