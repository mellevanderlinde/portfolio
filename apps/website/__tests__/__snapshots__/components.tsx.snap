// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Match footer 1`] = `
<div>
  <small
    class="block lg:mt-24 mt-16 text-[#1C1C1C] dark:text-[#D4D4D4]"
  >
    <time>
      Â© 
      2024
    </time>
     
    Melle
    <div
      class="flex text-lg gap-3.5 float-right transition-opacity duration-300 hover:opacity-90"
    >
      <a
        href="https://github.com/mellevanderlinde"
        rel="noopener noreferrer"
        target="_blank"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 496 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"
          />
        </svg>
      </a>
      <a
        href="https://linkedin.com/in/mellevdlinde"
        rel="noopener noreferrer"
        target="_blank"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M100.28 448H7.4V148.9h92.88zM53.79 108.1C24.09 108.1 0 83.5 0 53.8a53.79 53.79 0 0 1 107.58 0c0 29.7-24.1 54.3-53.79 54.3zM447.9 448h-92.68V302.4c0-34.7-.7-79.2-48.29-79.2-48.29 0-55.69 37.7-55.69 76.7V448h-92.78V148.9h89.08v40.8h1.3c12.4-23.5 42.69-48.3 87.88-48.3 94 0 111.28 61.9 111.28 142.3V448z"
          />
        </svg>
      </a>
      <a
        href="https://github.com/mellevanderlinde/portfolio"
        rel="noopener noreferrer"
        target="_blank"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 640 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M392.8 1.2c-17-4.9-34.7 5-39.6 22l-128 448c-4.9 17 5 34.7 22 39.6s34.7-5 39.6-22l128-448c4.9-17-5-34.7-22-39.6zm80.6 120.1c-12.5 12.5-12.5 32.8 0 45.3L562.7 256l-89.4 89.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0l112-112c12.5-12.5 12.5-32.8 0-45.3l-112-112c-12.5-12.5-32.8-12.5-45.3 0zm-306.7 0c-12.5-12.5-32.8-12.5-45.3 0l-112 112c-12.5 12.5-12.5 32.8 0 45.3l112 112c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L77.3 256l89.4-89.4c12.5-12.5 12.5-32.8 0-45.3z"
          />
        </svg>
      </a>
    </div>
  </small>
</div>
`;

exports[`Match header 1`] = `
<div>
  <h1
    class="mb-8 text-2xl font-medium tracking-tight"
  >
    Test
  </h1>
</div>
`;

exports[`Match navbar 1`] = `
<div>
  <nav
    class="lg:mb-16 mb-12 py-5"
  >
    <div
      class="flex flex-col md:flex-row md:items-center justify-between"
    >
      <div
        class="flex items-center"
      >
        <a
          class="text-3xl font-semibold tracking-tight"
          href="/"
        >
          Melle
        </a>
      </div>
      <div
        class="flex flex-row gap-4 mt-6 md:mt-0 md:ml-auto items-center"
      >
        <a
          class="transition-all hover:text-neutral-800 dark:hover:text-neutral-200 flex align-middle relative"
          href="/blog"
        >
          Blog
        </a>
        <a
          class="transition-all hover:text-neutral-800 dark:hover:text-neutral-200 flex align-middle relative"
          href="/projects"
        >
          Projects
        </a>
        <a
          class="transition-all hover:text-neutral-800 dark:hover:text-neutral-200 flex align-middle relative"
          href="/work"
        >
          Work
        </a>
      </div>
    </div>
  </nav>
</div>
`;

exports[`Match posts 1`] = `
[
  {
    "content": "<Callout emoji="ðŸ’¡">
  This post is also featured on the [PostNL Engineering blog](https://medium.com/postnl-engineering/connect-github-actions-with-aws-using-openid-connect-bf1e11401b50) on Medium!
</Callout>

A common pattern is to invoke AWS actions from a GitHub Actions workflow. 
For example, a workflow can be used for [AWS CDK](https://docs.aws.amazon.com/cdk/v2/guide/home.html) deployments, invoking a Lambda function (e.g., for integration tests) or reading data from DynamoDB. 

Using [OpenID Connect (OIDC)](https://openid.net/developers/how-connect-works/) such AWS actions can be invoked by an IAM role.
The role trusts a specific GitHub repository or branch, giving granular control over how GitHub Actions can use AWS credentials.
This solution uses temporary credentials, meaning that there is no need to store and rotate long-lived credentials.

This post demonstrates how to allow a specific GitHub repository branch to access AWS (with AWS CDK in TypeScript) and invoke actions on an AWS account (with GitHub Actions).

The code used in this post can be found on [GitHub](https://github.com/mellevanderlinde/oidc-aws-github).

## Create IAM OIDC Identity Provider

To establish a trust relationship between GitHub and AWS, create an [IAM OIDC Identity Provider](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_create_oidc.html) for GitHub.

\`\`\`ts
const oidcProvider = new iam.CfnOIDCProvider(this, "GithubOidcProvider", {
  url: "https://token.actions.githubusercontent.com",
  clientIdList: ["sts.amazonaws.com"],
});
\`\`\`

## Create IAM Role

Next, the created identity provider is used to create an IAM role. 
The trust policy of the role defines which specific GitHub repository branch is allowed to assume the role (using the pattern \`repo:OWNER/REPOSITORY:ref:refs/heads/BRANCH\`).

\`\`\`ts
const conditions: iam.Conditions = {
  StringEquals: {
    "token.actions.githubusercontent.com:aud": "sts.amazonaws.com",
    "token.actions.githubusercontent.com:sub":
      "repo:mellevanderlinde/oidc-aws-github:ref:refs/heads/main",
  },
};

const oidcRole = new iam.Role(this, "GithubOidcRole", {
  roleName: "github-oidc-role",
  assumedBy: new iam.WebIdentityPrincipal(oidcProvider.ref, conditions),
});
\`\`\`

Alternatively, an entire GitHub repository can be allowed using a wildcard \`*\` in the \`conditions\`. For this, \`StringLike\` is required instead of \`StringEquals\`.

\`\`\`ts
const conditions: iam.Conditions = {
  StringEquals: {
    "token.actions.githubusercontent.com:aud": "sts.amazonaws.com",
  },
  StringLike: {
    "token.actions.githubusercontent.com:sub":
      "repo:mellevanderlinde/oidc-aws-github:*",
  },
};
\`\`\`

For other patterns (e.g. allowing specific GitHub environments), see [Example subject claims](https://docs.github.com/en/actions/security-for-github-actions/security-hardening-your-deployments/about-security-hardening-with-openid-connect#example-subject-claims).

## Attach Permissions to IAM Role

To invoke AWS actions, the OIDC role needs to have the correct permissions. 
In this case the \`cdk diff\` command will be run, requiring read only permissions.
Therefore, the OIDC role is granted permission to assume the lookup role (created when the AWS account is [CDK bootstrapped](https://docs.aws.amazon.com/cdk/v2/guide/bootstrapping.html)).

\`\`\`ts
const lookupRole = iam.Role.fromRoleName(
  this,
  "LookupRole",
  \`cdk-hnb659fds-lookup-role-\${this.account}-\${this.region}\`,
);

lookupRole.grantAssumeRole(oidcRole);
\`\`\`

## Assume IAM Role from GitHub Actions Workflow

Next, the IAM role is assumed from a GitHub Actions workflow. 
This is done with the action [configure-aws-credentials](https://github.com/aws-actions/configure-aws-credentials), resulting in the following \`.github/workflows/oidc.yml\` file.
Note that a dummy AWS account ID is used.

\`\`\`yaml
on: push

jobs:
  oidc:
    permissions:
      id-token: write
      contents: read
    runs-on: ubuntu-latest
    steps:
      - name: Checkout code
        uses: actions/checkout@v4

      - name: Assume OIDC role
        uses: aws-actions/configure-aws-credentials@v4
        with:
          aws-region: eu-west-1
          role-to-assume: arn:aws:iam::012345678912:role/github-oidc-role
\`\`\`

## Invoke AWS Actions from GitHub Actions Workflow

The last step is to invoke AWS actions from the workflow. In this case, the \`cdk diff\` command is run by adding the following step.

\`\`\`yaml
- name: Invoke AWS action
  run: |
    npm ci
    npx cdk diff
\`\`\`

This results in the following \`.github/workflows/oidc.yml\` file.

\`\`\`yaml
on: push

jobs:
  oidc:
    permissions:
      id-token: write
      contents: read
    runs-on: ubuntu-latest
    steps:
      - name: Checkout code
        uses: actions/checkout@v4

      - name: Assume OIDC role
        uses: aws-actions/configure-aws-credentials@v4
        with:
          aws-region: eu-west-1
          role-to-assume: arn:aws:iam::012345678912:role/github-oidc-role
          
      - name: Invoke AWS action
        run: |
          npm ci
          npx cdk diff
\`\`\`

## Conclusion

Using OpenID Connect, integrating AWS with GitHub Actions is relatively simple and secure.
This can be useful for deployments, testing purposes or other cases where AWS actions need to be invoked.

## Resources

- [Example repository with OIDC for AWS and GitHub](https://github.com/mellevanderlinde/oidc-aws-github)
- [Configuring OpenID Connect in Amazon Web Services](https://docs.github.com/en/actions/security-for-github-actions/security-hardening-your-deployments/configuring-openid-connect-in-amazon-web-services)
- [Use IAM roles to connect GitHub Actions to actions in AWS](https://aws.amazon.com/blogs/security/use-iam-roles-to-connect-github-actions-to-actions-in-aws/)
- [About security hardening with OpenID Connect](https://docs.github.com/en/actions/security-for-github-actions/security-hardening-your-deployments/about-security-hardening-with-openid-connect)",
    "metadata": {
      "publishedAt": "2024-09-26",
      "title": "Connect GitHub Actions with AWS using OpenID Connect",
    },
    "slug": "openid-connect-aws-github",
  },
]
`;

exports[`Match robots 1`] = `
{
  "rules": [
    {
      "userAgent": "*",
    },
  ],
  "sitemap": "https://mellevanderlinde.com/sitemap.xml",
}
`;

exports[`Match sitemap 1`] = `
[
  {
    "lastModified": "2024-01-01T00:00:00.000Z",
    "url": "https://mellevanderlinde.com/",
  },
  {
    "lastModified": "2024-01-01T00:00:00.000Z",
    "url": "https://mellevanderlinde.com/blog",
  },
  {
    "lastModified": "2024-01-01T00:00:00.000Z",
    "url": "https://mellevanderlinde.com/projects",
  },
  {
    "lastModified": "2024-01-01T00:00:00.000Z",
    "url": "https://mellevanderlinde.com/work",
  },
  {
    "lastModified": "2024-01-01T00:00:00.000Z",
    "url": "https://mellevanderlinde.com/blog/openid-connect-aws-github",
  },
]
`;
